app/designer/features/props-panel/use-props-logic.ts

```diff
@@ -0,0 +1,169 @@
+import type {
+  DeleteProp,
+  Publish,
+  UserProp,
+  UserPropsUpdates,
+} from "@webstudio-is/sdk";
+import { componentsMeta } from "@webstudio-is/sdk";
+import ObjectId from "bson-objectid";
+import produce from "immer";
+import uniqBy from "lodash/uniqBy";
+import debounce from "lodash/debounce";
+import { useRef, useState } from "react";
+import type { SelectedInstanceData } from "~/shared/canvas-components";
+
+export type handleChangePropType = (
+  id: UserProp["id"],
+  field: "prop" | "value",
+  value: UserProp["prop"] | UserProp["value"]
+) => void;
+
+const getRequiredProps = (
+  selectedInstanceData: SelectedInstanceData
+): UserProp[] => {
+  const { component } = selectedInstanceData;
+  const meta = componentsMeta[component];
+  const argTypes = meta?.argTypes || {};
+  return Object.entries(argTypes)
+    .filter(([_, value]) => value.required)
+    .map(([prop, _]) => ({
+      id: ObjectId().toString(),
+      prop,
+      value: "",
+      required: true,
+    }));
+};
+
+// TODO: This returns same props for all instances.
+// See the failing test in use-props-logic.test.ts
+const getPropsWithDefaultValue = (
+  selectedInstanceData: SelectedInstanceData
+): UserProp[] => {
+  const { component } = selectedInstanceData;
+  const meta = componentsMeta[component];
+  const argTypes = meta?.argTypes || {};
+  return Object.entries(argTypes)
+    .filter(([_, value]) => value.defaultValue != null)
+    .map(([prop, propObj]) => {
+      const { defaultValue } = propObj;
+      const value = "value" in defaultValue ? defaultValue.value : defaultValue;
+      return {
+        id: ObjectId().toString(),
+        prop,
+        value,
+      };
+    });
+};
+
+type UsePropsLogic = {
+  publish: Publish;
+  selectedInstanceData: SelectedInstanceData;
+};
+
+export const usePropsLogic = ({
+  selectedInstanceData,
+  publish,
+}: UsePropsLogic) => {
+  const props =
+    selectedInstanceData.props === undefined ||
+    selectedInstanceData.props.props.length === 0
+      ? []
+      : selectedInstanceData.props.props;
+
+  const initialState = uniqBy(
```

Comment:
the amount of data bassed to this function is not making it a candidate for memoization?

Reply:
No, this is a premature optimization at its best. The amount of data = max dozens of objects.

Reply:
dozens is fine, I was thinking this could be all of the props, but I guess in this case its only a few indeed